### PR TITLE
KafkaTestCluster - Extend ability to create a cluster of 1 or more kafka brokers.

### PR DESCRIPTION
**Pull request by Crim**
Adds the ability to run an entire 'cluster' of kafka brokers.  Doing so will allow the ability to easily write tests that validate functionality when:

* Multiple brokers are required.
* Handling when broker's become unavailable.
* ISR or Leader changes.

### CODE REVIEW

File: kafka-junit-core/src/main/java/com/salesforce/kafka/test/KafkaCluster.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+     */
+    String getKafkaConnectString();
+
+    /**
+     * @return The proper connect string to use for Zookeeper.
+     */
+    String getZookeeperConnectString();
+
+    // TODO: From here down provides backwards compatibility.  Maybe we just want to instead add a getter for
+    // an instance of KafkaTestUtils and allow access to these methods only via it?
```
Comment (stanlemon) [inline kafka-junit-core/src/main/java/com/salesforce/kafka/test/KafkaCluster.java]: I think this is a better route.

File: kafka-junit-core/src/main/java/com/salesforce/kafka/test/KafkaTestServer.java [hunk 2]
```diff
@@ -31,29 +31,21 @@
...
-import java.util.concurrent.ExecutionException;
 
 /**
  * This will spin up a ZooKeeper and Kafka server for use in integration tests. Simply
  * create an instance of KafkaTestServer and call start() and you can publish to Kafka
  * topics in an integration test. Be sure to call shutdown() when the test is complete
  * or use the AutoCloseable interface.
  */
-public class KafkaTestServer implements AutoCloseable {
+public class KafkaTestServer implements KafkaCluster, KafkaProvider {
```
Comment (stanlemon) [inline kafka-junit-core/src/main/java/com/salesforce/kafka/test/KafkaTestServer.java]: I would leave AutoCloseable, just for legibility reasons.

File: kafka-junit-core/src/main/java/com/salesforce/kafka/test/KafkaTestServer.java [hunk 3]
```diff
@@ -114,6 +111,23 @@ public KafkaTestServer(final Properties overrideBrokerProperties) {
         this.overrideBrokerProperties.putAll(overrideBrokerProperties);
     }
 
+    /**
+     * Package protected constructor allowing override of ZookeeperTestServer instance.
```
Comment (stanlemon) [inline kafka-junit-core/src/main/java/com/salesforce/kafka/test/KafkaTestServer.java]: What’s the use case? Just tests?

File: kafka-junit-core/src/main/java/com/salesforce/kafka/test/KafkaTestUtils.java [hunk 4]
```diff
@@ -172,37 +191,148 @@ public KafkaTestUtils(KafkaTestServer kafkaTestServer) {
...
+
+    /**
+     * Creates a topic in Kafka. If the topic already exists this does nothing.
+     * @param topicName the namespace name to create.
+     * @param partitions the number of partitions to create.
+     * @param replicationFactor the number of replicas for the topic.
+     */
+    public void createTopic(final String topicName, final int partitions, final short replicationFactor) {
+        // Create admin client
+        try (final AdminClient adminClient = getAdminClient()) {
```
Comment (stanlemon) [inline kafka-junit-core/src/main/java/com/salesforce/kafka/test/KafkaTestUtils.java]: This doesn’t look right.

File: kafka-junit-core/src/main/java/com/salesforce/kafka/test/KafkaTestUtils.java [hunk 5]
```diff
@@ -172,37 +191,148 @@ public KafkaTestUtils(KafkaTestServer kafkaTestServer) {
...
+     * @param <V> Type of message value
+     * @param keySerializer Class of serializer to be used for keys.
+     * @param valueSerializer Class of serializer to be used for values.
+     * @param config Additional producer configuration options to be set.
+     * @return KafkaProducer configured to produce into Test server.
+     */
+    public <K, V> KafkaProducer<K, V> getKafkaProducer(
+        final Class<? extends Serializer<K>> keySerializer,
+        final Class<? extends Serializer<V>> valueSerializer,
+        final Properties config) {
```
Comment (stanlemon) [inline kafka-junit-core/src/main/java/com/salesforce/kafka/test/KafkaTestUtils.java]: Need a new line here.

File: kafka-junit-core/src/test/java/com/salesforce/kafka/test/KafkaTestServerTest.java [hunk 6]
```diff
@@ -28,6 +28,8 @@
 import com.google.common.collect.Iterables;
 import org.apache.kafka.clients.admin.AdminClient;
 import org.apache.kafka.clients.admin.DescribeClusterResult;
+import org.apache.kafka.clients.admin.NewTopic;
```
Comment (stanlemon) [inline kafka-junit-core/src/test/java/com/salesforce/kafka/test/KafkaTestServerTest.java]: Are these new imports used?

File: kafka-junit-core/src/main/java/com/salesforce/kafka/test/KafkaBrokerList.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    /**
+     * Constructor.
+     * @param brokers List of KafkaBrokers in a cluster.
+     */
+    public KafkaBrokerList(final List<KafkaBroker> brokers) {
+        // Create new map.
+        final Map<Integer, KafkaBroker> map = new HashMap<>();
+
+        // Add entries to the map.
+        brokers.forEach((broker) -> {
```
[the remark this answers is not in the document]
Comment (Crim) [reply kafka-junit-core/src/main/java/com/salesforce/kafka/test/KafkaBrokerList.java]: Nice! Learn something new every day :)

**Comment by Crim** [comment]
Thanks for your input @Cs4r ! Definitely appreciated!  

I think the KafkaBroker class needs to be modified to return the actual KafkaTestServer instance, or at least contain methods allowing for starting/stopping each individual broker.  Doing so will allow for testing things like how applications handle Leadership changes, or brokers just becoming unavailable, which at the moment isn't actually possible.

**Pull request merged**